### PR TITLE
Correctly re-instate Matplotlib converters

### LINKED ISSUE #27479: deregister_matplotlib_converters() in 0.25 incorrectly removes Matplotlib converters
Consider

```python
import matplotlib.units as munits
import pandas as pd
pd.plotting.deregister_matplotlib_converters()

print(pd.__version__)
for key in munits.registry:
    print(key, '\t', munits.registry[key])
```

`deregister_matplotlib_converters()` should only remove `pandas` converters, and not any other converters in the MPL units registry, but in pandas 0.25 it removes some of the Matplotlib converters.

#### Problem description

For pandas 0.24 this prints:
```
0.24.2
<class 'decimal.Decimal'> 	 <matplotlib.units.DecimalConverter object at 0x1135515f8>
<class 'datetime.datetime'> 	 <matplotlib.dates.DateConverter object at 0x11ed79cc0>
<class 'datetime.date'> 	 <matplotlib.dates.DateConverter object at 0x11ed79c88>
<class 'numpy.datetime64'> 	 <matplotlib.dates.DateConverter object at 0x11ed79c50>
```
But for pandas 0.25 it gives
```
0.25.0
<class 'decimal.Decimal'> 	 <matplotlib.units.DecimalConverter object at 0x1133cd588>
```

This means `pandas.plotting.deregister_matplotlib_converters` is removing some of the built in matplotlib converters from the units registry when it shouldn't be; this is causing some of our tests to break upstream (xref https://github.com/matplotlib/matplotlib/issues/14859)

### PR DESCRIPTION
**Pull request by dstansby**
Fixes https://github.com/pandas-dev/pandas/issues/27479. Converters should only be cached to be re-instated if they are original, and not pandas converters.

I'm struggling to write a test for this; I think it requires a clean state with no pandas imports, so one can check what was in the units registry before pandas was imported. I think this means a new test file with no pandas imports at the top, but maybe someone has a better idea?

### CODE REVIEW

**Comment by TomAugspurger** [comment]
Thanks for this!

> I'm struggling to write a test for this; 

Testing these is quite hard (we accidentally broke the warning for a release and didn't notice).

https://github.com/pandas-dev/pandas/blob/9bab81e013af33084c48b54dc2a6ef188acfd225/pandas/tests/plotting/test_converter.py#L31 works reasonably well I think. The idea is to run the check in a subprocess.

**Comment by TomAugspurger** [comment]
A release note in whatsnew/0.25.1.rst would be nice to have. I think there's a "plotting" subsection.

**Comment by dstansby** [comment]
Thanks for the test tip, the new test fails before the changes in this PR.

File: doc/source/whatsnew/v0.25.1.rst [hunk 1]
```diff
@@ -113,7 +113,8 @@ I/O
 Plotting
 ^^^^^^^^
 
--
+- Fixed the re-instatement of Matplotlib datetime converters after calling
```
Comment (jreback) [inline doc/source/whatsnew/v0.25.1.rst]: can you add the issue number

File: pandas/plotting/_matplotlib/converter.py [hunk 2]
```diff
@@ -64,11 +64,12 @@ def register(explicit=True):
 
     pairs = get_pairs()
     for type_, cls in pairs:
-        converter = cls()
-        if type_ in units.registry:
+        # Cache previous converter if present
+        if type_ in units.registry and type(units.registry[type_]) is not cls:
```
Comment (jreback) [inline pandas/plotting/_matplotlib/converter.py]: can you use isinstance here

File: pandas/tests/plotting/test_converter.py [hunk 3]
```diff
@@ -40,6 +40,21 @@ def test_initial_warning():
     assert "Using an implicitly" in out
 
 
+def test_registry_mpl_resets():
+    # Check that Matplotlib converters are properly reset
```
Comment (jreback) [inline pandas/tests/plotting/test_converter.py]: can you add the issue number as a comment here

**Review by jreback — CHANGES_REQUESTED** [review]

File: doc/source/whatsnew/v0.25.1.rst [hunk 4]
```diff
@@ -113,7 +113,8 @@ I/O
...
+  `pandas.plotting.deregister_matplotlib_converters()` (issue #27481).
```
Comment (jreback) [inline doc/source/whatsnew/v0.25.1.rst]: sorry this needs to be like ``:issue:`27481` `` to render

**Review by jreback — CHANGES_REQUESTED** [review]
also pls merge master

**Comment by dstansby** [comment]
I'm away for the next week or so without a computer, so if anyone wants this in quicker than that feel free to force push to my branch.

**Comment by dstansby** [comment]
Is there anything else (apart from a rebase) I need to do on this?

**Review by TomAugspurger — APPROVED** [review]
Looks good. Ping on green.

**Comment by TomAugspurger** [comment]
Did you merge master before pushing? We've seen similar CI failures, but I thought we fixed them all on master.

**Comment by dstansby** [comment]
Yeah, I should have rebased on to master, I've done another rebase to see if that fixes the tests.

**Comment by TomAugspurger** [comment]
Merged master again to see if we can get the CI passing.

**Review by TomAugspurger — APPROVED** [review]

**Comment by TomAugspurger** [comment]
Thanks @dstansby. Sorry about the CI issues.

**Pull request merged**